usr.bin/at/at.1
.\" $NetBSD: at.1,v 1.26.2.1 2017/01/07 08:56:57 pgoyette Exp $
.\" $OpenBSD: at.1,v 1.6 1998/06/05 00:47:46 deraadt Exp $
.\" $FreeBSD: at.man,v 1.6 1997/02/22 19:54:05 peter Exp $
.Dd March 10, 2008
.Dt AT 1
.Os
.Sh NAME
.Nm at ,
.Nm batch ,
.Nm atq ,
.Nm atrm
.Nd queue, examine or delete jobs for later execution
.Sh SYNOPSIS
.Nm at
.Op Fl bdlmrVv
.Op Fl f Ar file
.Op Fl q Ar queue
.Fl t Ar [[CC]YY]MMDDhhmm[.SS]
.Nm
.Op Fl bdlmrVv
.Op Fl f Ar file
.Op Fl q Ar queue
.Ar time
.Nm
.Op Fl V
.Fl c Ar job Op Ar job ...
.Nm atq
.Op Fl Vv
.Op Fl q Ar queue
.Nm atrm
.Op Fl V
.Ar job
.Op Ar job ...
.Nm batch
.Op Fl mVv
.Op Fl f Ar file
.Op Fl q Ar queue
.Op Fl t Ar [[CC]YY]MMDDhhmm[.SS]
.Nm batch
.Op Fl mVv
.Op Fl f Ar file
.Op Fl q Ar queue
.Op Ar time
.Sh DESCRIPTION
.Nm
and
.Nm batch
read commands from standard input or a specified file which
are to be executed at a later time, using
.Xr sh 1 .
.Bl -tag -width indent
.It Nm at
Executes commands at a specified time.
.It Nm atq
Lists the user's pending jobs, unless the user is the superuser.
In that case, everybody's jobs are listed.
.It Nm atrm
Deletes jobs.
.It Nm batch
Executes commands when system load levels permit.
In other words, when
the load average drops below 1.5, or the value specified in the invocation of
.Xr atrun 8 .
.El
.Pp
.Nm
allows some moderately complex
.Ar time
specifications.
It accepts times of the form
.Ar HHMM
or
.Ar HH:MM
to run a job at a specific time of day.
(If that time is already past, the next day is assumed.)
You may also specify
.Sq midnight ,
.Sq noon ,
or
.Sq teatime
(4pm)
and you can have a time-of-day suffixed with
.Sq AM
or
.Sq PM
for running in the morning or the evening.
You can also say what day the job will be run,
by giving a date in the form
.Ar %month-name day
with an optional
.Ar year ,
or giving a date of the form
.Ar MMDDYY
or
.Ar MM/DD/YY
or
.Ar DD.MM.YY .
The specification of a date must follow the specification of
the time of day.
You can also give times like
.Op Nm now
or
.Op Nm now
.Sq + Ar count %time-units ,
where the time-units can be
.Sq minutes ,
.Sq hours ,
.Sq days ,
.Sq weeks ,
.Sq months ,
or
.Sq years
and you can tell
.Nm
to run the job today by suffixing the time with
.Sq today
and to run the job tomorrow by suffixing the time with
.Sq tomorrow .
.Pp
For example, to run a job at 4pm three days from now, you would do
.Dl at 4pm + 3 days ,
to run a job at 10:00am on July 31, you would do
.Dl at 10am Jul 31
and to run a job at 1am tomorrow, you would do
.Dl at 1am tomorrow .
.Pp
Alternatively the time may be specified in a language-neutral fashion
by using the
.Fl t
options.
.Pp
For both
.Nm
and
.Nm batch ,
commands are read from standard input or the file specified
with the
.Fl f
option and executed.
The working directory, the environment (except for the variables
.Ev TERM ,
.Ev TERMCAP ,
.Ev DISPLAY
and
.Ev _ )
and the
.Ar umask
are retained from the time of invocation.
An
.Nm
or
.Nm batch
command invoked from a
.Xr su 1
shell will retain the current userid.
The user will be mailed standard error and standard output from his
commands, if any.
Mail will be sent using the command
.Xr sendmail 1 .
If
.Nm
is executed from a
.Xr su 1
shell, the owner of the login shell will receive the mail.
.Pp
The superuser may use these commands in any case.
For other users, permission to use
.Nm at
is determined by the files
.Pa /var/at/at.allow
and
.Pa /var/at/at.deny .
.Pp
If the file
.Pa /var/at/at.allow
exists, only usernames mentioned in it are allowed to use
.Nm .
.Pp
If
.Pa /var/at/at.allow
does not exist,
.Pa /var/at/at.deny
is checked; every username not mentioned in it is then allowed
to use
.Nm .
.Pp
If neither exists, only the superuser is allowed use of
.Nm .
.Pp
An empty
.Pa /var/at/at.deny
means that every user is allowed use these commands.
This is the default configuration.
.Sh OPTIONS
.Bl -tag -offset indent -width XqXqueueXX
.It Fl b
Is an alias for
.Nm batch .
.It Fl c
Cats the jobs listed on the command line to standard output.
.It Fl d
Is an alias for
.Nm atrm .
.It Fl f Ar file
Reads the job from
.Ar file
rather than standard input.
.It Fl l
Is an alias for
.Nm atq .
.It Fl m
Send mail to the user when the job has completed even if there was no
output.
.It Fl q Ar queue
Uses the specified queue.
A queue designation consists of a single letter.
Valid queue designations
range from
.Sq a
to
.Sq z
and
.Sq A
to
.Sq Z .
The
.Sq c
queue is the default for
.Nm
and the
.Sq E
queue for
.Nm batch .
Queues with higher letters run with increased niceness.
If a job is submitted to a queue designated with an uppercase letter, it
is treated as if it had been submitted to batch at that time.
If
.Nm atq
is given a specific queue, it will only show jobs pending in that queue.
.It Fl r
Is an alias for
.Nm atrm .
.It Fl t
For both
.Nm
and
.Nm batch ,
the time may be specified in a language-neutral format consisting of:
.Bl -tag -width Ds -compact -offset indent
.It Ar CC
The first two digits of the year (the century).
.It Ar YY
The second two digits of the year.
If
.Ar YY
is specified, but
.Ar CC
is not, a value for
.Ar YY
between 69 and 99 results in a
.Ar CC
value of 19.
Otherwise, a
.Ar CC
value of 20 is used.
.It Ar MM
The month of the year, from 01 to 12.
.It Ar \&DD
The day of the month, from 01 to 31.
.It Ar hh
The hour of the day, from 00 to 23.
.It Ar mm
The minute of the hour, from 00 to 59.
.It Ar \&SS
The second of the minute, from 00 to 61.
.El
.It Fl V
Prints the version number to standard error.
.It Fl v
For
.Nm atq ,
shows completed but not yet deleted jobs in the queue.
Otherwise shows the time the job will be executed.
.El
.Sh FILES
.Bl -tag -width /var/at/.lockfile -compact
.It Pa /var/at/jobs
Directory containing job files
.It Pa /var/at/spool
Directory containing output spool files
.It Pa /var/run/utmp
Login records
.It Pa /var/at/at.allow
Allow permission control
.It Pa /var/at/at.deny
Deny permission control
.It Pa /var/at/.lockfile
Job-creation lock file.
.El
.Sh SEE ALSO
.Xr nice 1 ,
.Xr sendmail 1 ,
.Xr sh 1 ,
.Xr umask 2 ,
.Xr atrun 8 ,
.Xr cron 8
.Sh STANDARDS
The
.Nm
and
.Nm batch
utilities conform to
.St -p1003.2-92 .
.Sh AUTHORS
.An -nosplit
.Nm
was mostly written by
.An Thomas Koenig Aq Mt ig25@rz.uni-karlsruhe.de .
The time parsing routines are implemented by
.An David Parsons Aq Mt orc@pell.chi.il.us .
.Sh BUGS
If the file
.Pa /var/run/utmp
is not available or corrupted, or if the user is not logged on at the
time
.Nm
is invoked, the mail is sent to the userid found
in the environment variable
.Ev LOGNAME .
If that is undefined or empty, the current userid is assumed.
.Pp
.Nm
and
.Nm batch
as presently implemented are not suitable when users are competing for
resources.
If this is the case for your site, you might want to consider another
batch system, such as
.Ic nqs .


doc/CHANGES
# LIST OF CHANGES FROM LAST RELEASE:			<$Revision: 1.2237 $>
#
#
# [Note: This file does not mention every change made to the NetBSD source tree.
# For an exhaustive list, please subscribe to the `source-changes' mailing list,
# or see the mailing list archives.  For more information, send a message
# containing just the word `help' to `majordomo@NetBSD.org'.]
#
# ----------------------------------------------------------------------------
# PLEASE USE THE FOLLOWING STYLE WHEN ADDING ITEMS:
#
#	theme: Content ...
#		more content ...
#		more content. [developer DATE-FORMAT]
#
# That's it, one tab to start an item, and two tabs to indent the content.
# Also please start the content with a capital and end it with a dot.
#
# This is very important for the www scripts, otherwise it won't be able to
# parse the file and the HTML output will be mangled.
#
# See htutils/changes/changes2html script for more details.
#
# ----------------------------------------------------------------------------


Changes from NetBSD 7.0 to NetBSD 8.0:
	libc: Update to tzcode2014f. [christos 20140815]
	gmake:	Update to gmake-3.81 in external/gpl2/gmake and remove 
		gnu/dist/gmake (3.80) [christos 20140818]
	module(7): Create MODULAR_DEFAULT_AUTOLOAD config flag to control
		whether module autoloading is on by default.
		[jnemeth 20140824]
	kernel: Option SYMTAB_SPACE has been replaced with COPY_SYMTAB.
		The related changes to src/sys/conf require a rebuild of
		config(1). [joerg 20140824]
	zoneinfo: Import tzdata2014g. [apb 20140831]
	wm(4): Add internal SERDES support for 82575 and newer controllers.
		[msaitoh 20140903]
	or1k: Add new port.  [matt 20140903]
	awindma(4): Add support for Allwinner A10/A20 DMA controller.
		[jmcneill 20140906]
	awinac(4): Add support for Allwinner A10/A20 audio codec.
		[jmcneill 20140906]
	awinrtc(4): Add support for Allwinner A10/A20 real-time clock.
		[jmcneill 20140907]
	awinmmc(4): Use DMA for MMC transfers. [jmcneill 20140908]
	axp20x(4): Add AXP20x PMU driver. [jmcneill 20140909]
	rtsol(8): Removed in favour of dhcpcd. [roy 20140911]
	rtsold(8): Removed in favour of dhcpcd. [roy 20140911]
	dhcpcd(8): Import dhcpcd-6.4.5. [roy 20140918]
	arm: Add support for i.MX6 SoC. [ryo 20140925]
	gpt(8): Completed overhaul, including adding follwing subcommands:
		resize, set, unset, backup, restore, and resizedisk.
		[jnemeth 20140926]
	dhcpcd(8): Import dhcpcd-6.4.7. [roy 20140927]
	mpt(4): added bio(4) support. [jmcneill 20140927]
	gpt(8): added type subcommand to modify partition type.
		[jnemeth 20140927]
	zoneinfo: Import tzdata2014h. [apb 20141004]
	dhcpcd(8): Import dhcpcd-6.5.0. [roy 20141006]
	libc: Update to tzcode2014h. [christos 20141007]
	file(1): Upgraded to 5.20. [christos 20141010]
	arm: Add support for Allwinner A31 SoC. [jmcneill 20141010]
	awiniic(4): Add support for Allwinner A31 SoC. [jmcneill 20141012]
	awinp2wi(4): Add support for Allwinner A31 Push-Pull Two Wire
		Interface. [jmcneill 20141012]
	axp22x(4): Add AXP22x PMU driver. [jmcneill 20141012]
	awinrtc(4): Add support for Allwinner A31 real-time clock.
		[jmcneill 20141012]
	awindma(4): Add support for Allwinner A31 DMA controller.
		[jmcneill 20141013]
	awinac(4): Add support for Allwinner A31 audio codec.
		[jmcneill 20141013]
	motg(4): Add support for Allwinner A10/A20 SoC. [jmcneill 20141015]
	wpa: Import wpa_supplicant and hostapd 2.3. [christos 20141016]
	dhcpcd(8): Import dhcpcd-6.5.1. [roy 20141018]
	OpenSSH: Imported 6.7. [christos 20141018]
	resolvconf(8): Import openresolv-3.6.0 [roy 20141020]
	gcc: Support -fsanitize=address [christos 20141022]
	zoneinfo: Import tzdata2014i. [apb 20141023]
	libc: Update to tzcode2014i. [christos 20141023]
	openpam(3): update to 20140912 (ourouparia) [christos 20141024]
	pppd(8): updated to version 2.4.7. [christos 20141025]
	acpi(4): Updated ACPICA to 20140926. [christos 20141025]
	dhcpcd(8): Import dhcpcd-6.6.0 [roy 20141029]
	resolvconf(8): Import openresolv-3.6.1 [roy 20141029]
	rtsx(4): Add support for Realtek RTS5227/RTL8402/RTL8411/RTL8411B.
		[nonaka 20141029]
	flex(1): Import flex-2.5.37 [christos 20141029]
	iwn(4): Add support for Intel Centrino Wireless-N
		100/105/130/135/2200/2230. From OpenBSD. [nonaka 20141030]
	awge(4): Add support for gigabit ethernet found on various
		Allwinner SOCs. [martin 20141101]
	dhcpcd(8): Import dhcpcd-6.6.1. [roy 20141107]
	awinhdmi(4): Add support for Allwinner A20/A31 HDMI controller.
		[jmcneill 20141110]
	awintcon(4): Add support for Allwinner A20/A31 TV/LCD controller.
		[jmcneill 20141110]
	awindebe(4): Add support for Allwinner A20/A31 Display engine
		backend (DE-BE). [jmcneill 20141110]
	awinhdmiaudio(4): Add support for Allwinner A20/A31 HDMI audio
		controller. [jmcneill 20141111]
	dhcpcd(8): Import dhcpcd-6.6.2. [roy 20141114]
	awinir(4): Add support for Allwinner A20/A31 IR receiver.
		[jmcneill 20141115]
	zoneinfo: Import tzdata2014j. [apb 20141116]
	libc: Change arc4random(3) to use ChaCha20 and per-thread state.
		[riastradh 20141116]
	libc: Update to tzcode2014j. [christos 20141117]
	libpcap: Import 1.6.2. [christos 20141119]
	tcpdump(8): Import 4.6.2. [christos 20141119]
	dhcpcd(8): Import dhcpcd-6.6.4. [roy 20141126]
	dhcpcd(8): Import dhcpcd-6.6.5. [roy 20141209]
	bind: Import version 9.10.1-P1. [christos 20141209]
	dhcpcd(8): Import dhcpcd-6.6.6. [roy 20141217]
	powerpc: Switch to GCC 4.8. [matt 20141218]
	dhcpcd(8): Import dhcpcd-6.6.7. [roy 20141219]
	ntp: Import ntp 4.2.8. [christos 20141219]
	midirecord(1): Add program to record MIDI files.  [mrg 20141229]
	file: update to 5.22 security limits [christos 20150102]
	byacc: update to 20141128 for more bison compat [christos 20150103]
	mgx: driver for Southland Media Systems (now Quantum 3D) MGX SBus
		graphics devices [macallan 20150106]
	gcc: Import GCC 4.8.4.  [mrg 20150109]
	openssl: Import openssl 1.0.1k [spz 20150113]
	am-utils: Upgrade to 6.2; adds nfsv4, lustre (which we can't use)
		nfs3 rpc support for amfs [christos 20150117]

usr.bin/at/at.c
/*	$NetBSD: at.c,v 1.31 2016/03/13 00:32:09 dholland Exp $	*/

/*
 *  at.c : Put file into atrun queue
 *  Copyright (C) 1993, 1994  Thomas Koenig
 *
 *  Atrun & Atq modifications
 *  Copyright (C) 1993  David Parsons
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. The name of the author(s) may not be used to endorse or promote
 *    products derived from this software without specific prior written
 *    permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR(S) ``AS IS'' AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
 * IN NO EVENT SHALL THE AUTHOR(S) BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
 * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

/* System Headers */
#include <sys/types.h>
#include <sys/param.h>
#include <sys/stat.h>
#include <sys/wait.h>
#include <ctype.h>
#include <dirent.h>
#include <err.h>
#include <errno.h>
#include <fcntl.h>
#include <locale.h>
#include <pwd.h>
#include <signal.h>
#include <stdbool.h>
#include <stddef.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <unistd.h>
#include <util.h>

/* Local headers */
#include "at.h"
#include "panic.h"
#include "parsetime.h"
#include "perm.h"
#include "pathnames.h"
#include "stime.h"
#include "privs.h"

/* Macros */
#define ALARMC 10		/* Number of seconds to wait for timeout */

#define TIMESIZE 50

enum { ATQ, ATRM, AT, BATCH, CAT };	/* what program we want to run */

/* File scope variables */
#ifndef lint
#if 0
static char rcsid[] = "$OpenBSD: at.c,v 1.15 1998/06/03 16:20:26 deraadt Exp $";
#else
__RCSID("$NetBSD: at.c,v 1.31 2016/03/13 00:32:09 dholland Exp $");
#endif
#endif

const char *no_export[] = {"TERM", "TERMCAP", "DISPLAY", "_"};
static int send_mail = 0;

/* External variables */

extern char **environ;
bool fcreated = false;
char atfile[FILENAME_MAX];

char *atinput = NULL;		/* where to get input from */
unsigned char atqueue = 0;	/* which queue to examine for jobs (atq) */
char atverify = 0;		/* verify time instead of queuing job */

/* Function declarations */

__dead static void sigc	(int);
__dead static void alarmc	(int);
static char *cwdname	(void);
static int  nextjob	(void);
static void writefile	(time_t, unsigned char);
static void list_jobs	(void);
static void process_jobs (int, char **, int);

/* Signal catching functions */

/*ARGSUSED*/
static void
sigc(int signo)
{

	/* If a signal interrupts us, remove the spool file and exit. */
	if (fcreated) {
		privs_enter();
		(void)unlink(atfile);
		privs_exit();
	}
	(void)raise_default_signal(signo);
	exit(EXIT_FAILURE);
}

/*ARGSUSED*/
static void
alarmc(int signo)
{

	/* Time out after some seconds. */
	warnx("File locking timed out");
	sigc(signo);
}

/* Local functions */

static char *
cwdname(void)
{

	/*
	 * Read in the current directory; the name will be overwritten on
	 * subsequent calls.
	 */
	static char path[MAXPATHLEN];

	return getcwd(path, sizeof(path));
}

static int
nextjob(void)
{
	int jobno;
	FILE *fid;

	if ((fid = fopen(_PATH_SEQFILE, "r+")) != NULL) {
		if (fscanf(fid, "%5x", &jobno) == 1) {
			(void)rewind(fid);
			jobno = (1+jobno) % 0xfffff;	/* 2^20 jobs enough? */
			(void)fprintf(fid, "%05x\n", jobno);
		} else
			jobno = EOF;
		(void)fclose(fid);
		return jobno;
	} else if ((fid = fopen(_PATH_SEQFILE, "w")) != NULL) {
		(void)fprintf(fid, "%05x\n", jobno = 1);
		(void)fclose(fid);
		return 1;
	}
	return EOF;
}

static void
writefile(time_t runtimer, unsigned char queue)
{
	/*
	 * This does most of the work if at or batch are invoked for
	 * writing a job.
	 */
	int jobno;
	char *ap, *ppos;
	const char *mailname;
	struct passwd *pass_entry;
	struct stat statbuf;
	int fdes, lockdes, fd2;
	FILE *fp, *fpin;
	struct sigaction act;
	char **atenv;
	int ch;
	mode_t cmask;
	struct flock lock;

	(void)setlocale(LC_TIME, "");

	/*
	 * Install the signal handler for SIGINT; terminate after removing the
	 * spool file if necessary
	 */
	(void)memset(&act, 0, sizeof(act));
	act.sa_handler = sigc;
	(void)sigemptyset(&act.sa_mask);
	act.sa_flags = 0;

	(void)sigaction(SIGINT, &act, NULL);

	(void)strlcpy(atfile, _PATH_ATJOBS, sizeof(atfile));
	ppos = atfile + strlen(atfile);

	/*
	 * Loop over all possible file names for running something at this
	 * particular time, see if a file is there; the first empty slot at
	 * any particular time is used.  Lock the file _PATH_LOCKFILE first
	 * to make sure we're alone when doing this.
	 */

	privs_enter();

	if ((lockdes = open(_PATH_LOCKFILE, O_WRONLY | O_CREAT, S_IWUSR | S_IRUSR)) < 0)
		perr("Cannot open lockfile " _PATH_LOCKFILE);

	lock.l_type = F_WRLCK;
	lock.l_whence = SEEK_SET;
	lock.l_start = 0;
	lock.l_len = 0;

	act.sa_handler = alarmc;
	(void)sigemptyset(&act.sa_mask);
	act.sa_flags = 0;

	/*
	 * Set an alarm so a timeout occurs after ALARMC seconds, in case
	 * something is seriously broken.
	 */
	(void)sigaction(SIGALRM, &act, NULL);
	(void)alarm(ALARMC);
	(void)fcntl(lockdes, F_SETLKW, &lock);
	(void)alarm(0);

	if ((jobno = nextjob()) == EOF)
	    perr("Cannot generate job number");

	(void)snprintf(ppos, sizeof(atfile) - (ppos - atfile),
	    "%c%5x%8lx", queue, jobno, (unsigned long) (runtimer/60));

	for (ap = ppos; *ap != '\0'; ap++)
		if (*ap == ' ')
			*ap = '0';

	if (stat(atfile, &statbuf) == -1)
		if (errno != ENOENT)
			perr("Cannot access " _PATH_ATJOBS);

	/*
	 * Create the file. The x bit is only going to be set after it has
	 * been completely written out, to make sure it is not executed in
	 * the meantime.  To make sure they do not get deleted, turn off
	 * their r bit.  Yes, this is a kluge.
	 */
	cmask = umask(S_IRUSR | S_IWUSR | S_IXUSR);
	if ((fdes = open(atfile, O_CREAT | O_TRUNC | O_WRONLY, S_IRUSR)) == -1)
		perr("Cannot create atjob file");

	if ((fd2 = dup(fdes)) == -1)
		perr("Error in dup() of job file");

	if (fchown(fd2, real_uid, real_gid) == -1)
		perr("Cannot give away file");

	privs_exit();

	/*
	 * We've successfully created the file; let's set the flag so it
	 * gets removed in case of an interrupt or error.
	 */
	fcreated = true;

	/* Now we can release the lock, so other people can access it */
	lock.l_type = F_UNLCK;
	lock.l_whence = SEEK_SET;
	lock.l_start = 0;
	lock.l_len = 0;
	(void)fcntl(lockdes, F_SETLKW, &lock);
	(void)close(lockdes);

	if ((fp = fdopen(fdes, "w")) == NULL)
		panic("Cannot reopen atjob file");

	/*
	 * Get the userid to mail to, first by trying getlogin(), which reads
	 * /etc/utmp, then from $LOGNAME or $USER, finally from getpwuid().
	 */
	mailname = getlogin();
	if (mailname == NULL && (mailname = getenv("LOGNAME")) == NULL)
		mailname = getenv("USER");

	if (mailname == NULL || mailname[0] == '\0' ||
	    strlen(mailname) > LOGIN_NAME_MAX || getpwnam(mailname) == NULL) {
		pass_entry = getpwuid(real_uid);
		if (pass_entry != NULL)
			mailname = pass_entry->pw_name;
	}

	if (atinput != NULL) {
		fpin = freopen(atinput, "r", stdin);
		if (fpin == NULL)
			perr("Cannot open input file");
	}
	(void)fprintf(fp,
	    "#!/bin/sh\n"
	    "# atrun uid=%u gid=%u\n"
	    "# mail %s %d\n",
	    real_uid, real_gid, mailname, send_mail);

	/* Write out the umask at the time of invocation */
	(void)fprintf(fp, "umask %o\n", cmask);

	/*
	 * Write out the environment. Anything that may look like a special
	 * character to the shell is quoted, except for \n, which is done
	 * with a pair of "'s.  Dont't export the no_export list (such as
	 * TERM or DISPLAY) because we don't want these.
	 */
	for (atenv = environ; *atenv != NULL; atenv++) {
		int export = 1;
		char *eqp;

		eqp = strchr(*atenv, '=');
		if (eqp == NULL)
			eqp = *atenv;
		else {
			size_t i;

			for (i = 0; i < __arraycount(no_export); i++) {
				export = export &&
				    strncmp(*atenv, no_export[i],
					(size_t)(eqp - *atenv)) != 0;
			}
			eqp++;
		}

		if (export) {
			(void)fwrite(*atenv, sizeof(char),
			    (size_t)(eqp - *atenv), fp);
			for (ap = eqp; *ap != '\0'; ap++) {
				if (*ap == '\n')
					(void)fprintf(fp, "\"\n\"");
				else {
					if (!isalnum((unsigned char)*ap)) {
						switch (*ap) {
						case '%': case '/': case '{':
						case '[': case ']': case '=':
						case '}': case '@': case '+':
						case '#': case ',': case '.':
						case ':': case '-': case '_':
							break;
						default:
							(void)fputc('\\', fp);
							break;
						}
					}
					(void)fputc(*ap, fp);
				}
			}
			(void)fputs("; export ", fp);
			(void)fwrite(*atenv, sizeof(char),
			    (size_t)(eqp - *atenv - 1), fp);
			(void)fputc('\n', fp);
		}
	}
	/*
	 * Cd to the directory at the time and write out all the
	 * commands the user supplies from stdin.
	 */
	(void)fputs("cd ", fp);
	for (ap = cwdname(); *ap != '\0'; ap++) {
		if (*ap == '\n')
			(void)fprintf(fp, "\"\n\"");
		else {
			if (*ap != '/' && !isalnum((unsigned char)*ap))
				(void)fputc('\\', fp);

			(void)fputc(*ap, fp);
		}
	}
	/*
	 * Test cd's exit status: die if the original directory has been
	 * removed, become unreadable or whatever.
	 */
	(void)fprintf(fp,
	    " || {\n"
	    "\t echo 'Execution directory inaccessible' >&2\n"
	    "\t exit 1\n"
	    "}\n");

	if ((ch = getchar()) == EOF)
		panic("Input error");

	do {
		(void)fputc(ch, fp);
	} while ((ch = getchar()) != EOF);

	(void)fprintf(fp, "\n");
	if (ferror(fp))
		panic("Output error");

	if (ferror(stdin))
		panic("Input error");

	(void)fclose(fp);

 	privs_enter();

	/*
	 * Set the x bit so that we're ready to start executing
	 */
	if (fchmod(fd2, S_IRUSR | S_IWUSR | S_IXUSR) == -1)
		perr("Cannot give away file");

	privs_exit();

	(void)close(fd2);
	(void)fprintf(stderr,
	    "Job %d will be executed using /bin/sh\n", jobno);
}

static void
list_jobs(void)
{
	/*
	 * List all a user's jobs in the queue, by looping through
	 * _PATH_ATJOBS, or everybody's if we are root
	 */
	struct passwd *pw;
	DIR *spool;
	struct dirent *dirent;
	struct stat buf;
	struct tm runtime;
	unsigned long ctm;
	unsigned char queue;
	int jobno;
	time_t runtimer;
	char timestr[TIMESIZE];
	int first = 1;

	privs_enter();

	if (chdir(_PATH_ATJOBS) == -1)
		perr("Cannot change to " _PATH_ATJOBS);

	if ((spool = opendir(".")) == NULL)
		perr("Cannot open " _PATH_ATJOBS);

	/* Loop over every file in the directory */
	while ((dirent = readdir(spool)) != NULL) {
		if (stat(dirent->d_name, &buf) == -1)
			perr("Cannot stat in " _PATH_ATJOBS);

		/*
		 * See it's a regular file and has its x bit turned on and
		 * is the user's
		 */
		if (!S_ISREG(buf.st_mode)
		    || (buf.st_uid != real_uid && real_uid != 0)
		    || !(S_IXUSR & buf.st_mode || atverify))
			continue;

		if (sscanf(dirent->d_name, "%c%5x%8lx", &queue, &jobno, &ctm) != 3)
			continue;

		if (atqueue && queue != atqueue)
			continue;

		runtimer = 60 * (time_t)ctm;
		runtime = *localtime(&runtimer);
#if 1
		/*
		 * Provide a consistent date/time format instead of a
		 * locale-specific one that might have 2 digit years
		 */
		(void)strftime(timestr, TIMESIZE, "%T %F", &runtime);
#else
		(void)strftime(timestr, TIMESIZE, "%X %x", &runtime);
#endif
		if (first) {
			(void)printf("%-*s  %-*s  %-*s  %s\n",
			    (int)strlen(timestr), "Date",
			    LOGIN_NAME_MAX, "Owner",
			    7, "Queue",
			    "Job");
			first = 0;
		}
		pw = getpwuid(buf.st_uid);

		(void)printf("%s  %-*s  %c%-*s  %d\n",
		    timestr,
		    LOGIN_NAME_MAX, pw ? pw->pw_name : "???",
		    queue,
		    6, (S_IXUSR & buf.st_mode) ? "" : "(done)",
		    jobno);
	}
	(void)closedir(spool);
	privs_exit();
}

static void
process_jobs(int argc, char **argv, int what)
{
	/* Delete every argument (job - ID) given */
	int i;
	struct stat buf;
	DIR *spool;
	struct dirent *dirent;
	unsigned long ctm;
	unsigned char queue;
	int jobno;

	privs_enter();

	if (chdir(_PATH_ATJOBS) == -1)
		perr("Cannot change to " _PATH_ATJOBS);

	if ((spool = opendir(".")) == NULL)
		perr("Cannot open " _PATH_ATJOBS);

	privs_exit();

	/* Loop over every file in the directory */
	while((dirent = readdir(spool)) != NULL) {

		privs_enter();
		if (stat(dirent->d_name, &buf) == -1)
			perr("Cannot stat in " _PATH_ATJOBS);
		privs_exit();

		if (sscanf(dirent->d_name, "%c%5x%8lx", &queue, &jobno, &ctm) !=3)
			continue;

		for (i = optind; i < argc; i++) {
			if (atoi(argv[i]) == jobno) {
				if (buf.st_uid != real_uid && real_uid != 0)
					errx(EXIT_FAILURE,
					    "%s: Not owner", argv[i]);

				switch (what) {
				case ATRM:
					privs_enter();

					if (unlink(dirent->d_name) == -1)
						perr(dirent->d_name);

					privs_exit();
					break;

				case CAT: {
					FILE *fp;
					int ch;

					privs_enter();

					fp = fopen(dirent->d_name, "r");

					privs_exit();

					if (!fp)
						perr("Cannot open file");
					else {
						while((ch = getc(fp)) != EOF)
							(void)putchar(ch);
						(void)fclose(fp);
					}
				}
					break;

				default:
					errx(EXIT_FAILURE,
					    "Internal error, process_jobs = %d",
						what);
					break;
				}
			}
		}
	}
	(void)closedir(spool);
}

/* Global functions */

int
main(int argc, char **argv)
{
	int c;
	unsigned char queue = DEFAULT_AT_QUEUE;
	char queue_set = 0;
	char time_set = 0;
	char *pgm;

	int program = AT;			/* our default program */
	const char *options = "q:f:t:mvldbrVc";	/* default options for at */
	int disp_version = 0;
	time_t timer;

	privs_relinquish();

	/* Eat any leading paths */
	if ((pgm = strrchr(argv[0], '/')) == NULL)
		pgm = argv[0];
	else
		pgm++;

	/* find out what this program is supposed to do */
	if (strcmp(pgm, "atq") == 0) {
		program = ATQ;
		options = "q:vV";
	} else if (strcmp(pgm, "atrm") == 0) {
		program = ATRM;
		options = "V";
	} else if (strcmp(pgm, "batch") == 0) {
		program = BATCH;
		options = "f:q:t:mvV";
	}

	/* process whatever options we can process */
	opterr = 1;
	while ((c = getopt(argc, argv, options)) != -1) {
		switch (c) {
		case 'v':	/* verify time settings */
			atverify = 1;
			break;

		case 'm':	/* send mail when job is complete */
			send_mail = 1;
			break;

		case 'f':
			atinput = optarg;
			break;

		case 'q':	/* specify queue */
			if (strlen(optarg) > 1)
				usage();

			atqueue = queue = *optarg;
			if (!(islower(queue) || isupper(queue)))
				usage();

			queue_set = 1;
			break;
		case 't':	/* touch(1) date format */
			timer = stime(optarg);
			time_set = 1;
			break;

		case 'd':
		case 'r':
			if (program != AT)
				usage();

			program = ATRM;
			options = "V";
			break;

		case 'l':
			if (program != AT)
				usage();

			program = ATQ;
			options = "q:vV";
			break;

		case 'b':
			if (program != AT)
				usage();

			program = BATCH;
			options = "f:q:mvV";
			break;

		case 'V':
			disp_version = 1;
			break;

		case 'c':
			program = CAT;
			options = "";
			break;

		default:
			usage();
			break;
		}
	} /* end of options eating */

	if (disp_version)
		(void)fprintf(stderr, "%s version %.1f\n", pgm, AT_VERSION);

	if (!check_permission())
		errx(EXIT_FAILURE,
		    "You do not have permission to use %s.", pgm);

	/* select our program */
	switch (program) {
	case ATQ:
		if (optind != argc)
			usage();
		list_jobs();
		break;

	case ATRM:
	case CAT:
		if (optind == argc)
			usage();
		process_jobs(argc, argv, program);
		break;

	case AT:
		if (argc > optind) {
			/* -t and timespec argument are mutually exclusive */
			if (time_set) {
				usage();
				exit(EXIT_FAILURE);
			} else {
				timer = parsetime(argc, argv);
				time_set = 1;
			}
		}

		if (atverify) {
			struct tm *tm = localtime(&timer);
			(void)fprintf(stderr, "%s\n", asctime(tm));
		}
		writefile(timer, queue);
		break;

	case BATCH:
		if (queue_set)
			queue = toupper(queue);
		else
			queue = DEFAULT_BATCH_QUEUE;

		if (argc > optind) {
			/* -t and timespec argument are mutually exclusive */
			if (time_set) {
				usage();
				exit(EXIT_FAILURE);
			} else {
				timer = parsetime(argc, argv);
				time_set = 1;
			}
		} else if (!time_set)
			timer = time(NULL);

		if (atverify) {
			struct tm *tm = localtime(&timer);
			(void)fprintf(stderr, "%s\n", asctime(tm));
		}

		writefile(timer, queue);
		break;

	default:
		panic("Internal error");
		break;
	}
	return EXIT_SUCCESS;
}
